File: nodejs/fix-images.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');

const readFile = promisify(fs.readFile);
const rename = promisify(fs.rename);
const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);

/**
 * 检测图片文件的真实格式
 * @param {string} filePath 文件路径
 * @returns {string|null} 真实的文件格式 ('jpg', 'gif', 'png', 'webp' 等)
 */
async function detectImageFormat(filePath) {
    try {
        // 只读取文件的前20个字节，足够识别大部分图片格式
        const buffer = await readFile(filePath, { start: 0, end: 19 });
        
        // GIF 格式检测
        if (buffer[0] === 0x47 && buffer[1] === 0x49 && buffer[2] === 0x46) {
            return 'gif';
        }
        
        // JPEG 格式检测
        if (buffer[0] === 0xFF && buffer[1] === 0xD8 && buffer[2] === 0xFF) {
            return 'jpg';
        }
        
        // PNG 格式检测
        if (buffer[0] === 0x89 && buffer[1] === 0x50 && buffer[2] === 0x4E && buffer[3] === 0x47) {
            return 'png';
        }
        
        // WebP 格式检测
        if (buffer.toString('ascii', 0, 4) === 'RIFF' && buffer.toString('ascii', 8, 12) === 'WEBP') {
            return 'webp';
        }
        
        // BMP 格式检测
        if (buffer[0] === 0x42 && buffer[1] === 0x4D) {
            return 'bmp';
        }
        
        // TIFF 格式检测
        if ((buffer[0] === 0x49 && buffer[1] === 0x49 && buffer[2] === 0x2A && buffer[3] === 0x00) ||
            (buffer[0] === 0x4D && buffer[1] === 0x4D && buffer[2] === 0x00 && buffer[3] === 0x2A)) {
            return 'tiff';
        }
        
        return null;
    } catch (error) {
        console.error(`读取文件失败: ${filePath}`, error.message);
        return null;
    }
}

/**
 * 获取文件扩展名（不包含点）
 * @param {string} filePath 文件路径
 * @returns {string} 扩展名
 */
function getFileExtension(filePath) {
    return path.extname(filePath).toLowerCase().substring(1);
}

/**
 * 处理单个目录中的所有图片文件
 * @param {string} dirPath 目录路径
 * @returns {Promise<{processed: number, renamed: number, errors: number}>}
 */
async function processDirectory(dirPath) {
    const stats = { processed: 0, renamed: 0, errors: 0 };
    
    try {
        const files = await readdir(dirPath);
        
        // 并发处理文件，但限制并发数量以避免文件系统压力
        const batchSize = 10;
        for (let i = 0; i < files.length; i += batchSize) {
            const batch = files.slice(i, i + batchSize);
            await Promise.all(batch.map(async (file) => {
                const filePath = path.join(dirPath, file);
                
                try {
                    const fileStat = await stat(filePath);
                    if (!fileStat.isFile()) return;
                    
                    const currentExtension = getFileExtension(file);
                    
                    // 只处理常见的图片扩展名
                    const imageExtensions = ['jpg', 'jpeg', 'gif', 'png', 'webp', 'bmp', 'tiff'];
                    if (!imageExtensions.includes(currentExtension)) return;
                    
                    stats.processed++;
                    
                    const actualFormat = await detectImageFormat(filePath);
                    if (!actualFormat) {
                        console.warn(`无法检测格式: ${file}`);
                        return;
                    }
                    
                    // 需要重命名的情况
                    const needsRename = (currentExtension !== actualFormat) && 
                                       !(currentExtension === 'jpeg' && actualFormat === 'jpg') &&
                                       !(currentExtension === 'jpg' && actualFormat === 'jpg');
                    
                    if (needsRename) {
                        const newName = file.replace(/\.[^.]+$/, `.${actualFormat}`);
                        const newPath = path.join(dirPath, newName);
                        
                        // 检查目标文件是否已存在
                        try {
                            await stat(newPath);
                            console.warn(`目标文件已存在，跳过重命名: ${file} -> ${newName}`);
                            return;
                        } catch (error) {
                            // 目标文件不存在，可以重命名
                        }
                        
                        await rename(filePath, newPath);
                        console.log(`已重命名: ${file} -> ${newName} (${currentExtension} -> ${actualFormat})`);
                        stats.renamed++;
                    } else {
                        console.log(`格式正确: ${file} (${actualFormat})`);
                    }
                } catch (error) {
                    console.error(`处理文件失败: ${file}`, error.message);
                    stats.errors++;
                }
            }));
        }
    } catch (error) {
        console.error(`读取目录失败: ${dirPath}`, error.message);
        stats.errors++;
    }
    
    return stats;
}

/**
 * 主函数
 */
async function main() {
    const args = process.argv.slice(2);
    const targetDir = args[0] || './Ori';
    
    console.log(`开始处理目录: ${targetDir}`);
    console.log('正在检测和修复图片文件格式...\n');
    
    const startTime = Date.now();
    const stats = await processDirectory(targetDir);
    const endTime = Date.now();
    
    console.log('\n=== 处理完成 ===');
    console.log(`处理文件数: ${stats.processed}`);
    console.log(`重命名文件数: ${stats.renamed}`);
    console.log(`错误文件数: ${stats.errors}`);
    console.log(`用时: ${(endTime - startTime)}ms`);
}

// 如果直接运行此脚本
if (require.main === module) {
    main().catch(error => {
        console.error('程序执行失败:', error);
        process.exit(1);
    });
}

module.exports = { detectImageFormat, processDirectory };

File: python/fix-images.py
#!/usr/bin/env python3
"""
图片格式检测和修复工具 - Python版本
检测图片文件的真实格式，并修正文件扩展名
"""

import os
import sys
import time
import asyncio
import aiofiles
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor
from typing import Optional, Dict, Tuple

class ImageFormatDetector:
    """图片格式检测器"""
    
    # 文件魔法数字（文件头）
    MAGIC_NUMBERS = {
        'gif': [b'GIF87a', b'GIF89a'],
        'jpg': [b'\xff\xd8\xff'],
        'png': [b'\x89PNG\r\n\x1a\n'],
        'webp': [b'RIFF', b'WEBP'],  # WebP需要特殊处理
        'bmp': [b'BM'],
        'tiff': [b'II*\x00', b'MM\x00*'],
    }
    
    @staticmethod
    async def detect_format(file_path: str) -> Optional[str]:
        """
        检测图片文件的真实格式
        
        Args:
            file_path: 文件路径
            
        Returns:
            真实的文件格式，如果无法识别则返回None
        """
        try:
            async with aiofiles.open(file_path, 'rb') as file:
                # 读取前20个字节，足够识别大部分图片格式
                header = await file.read(20)
                
                if len(header) < 4:
                    return None
                
                # GIF 检测
                if header.startswith(b'GIF87a') or header.startswith(b'GIF89a'):
                    return 'gif'
                
                # JPEG 检测
                if header.startswith(b'\xff\xd8\xff'):
                    return 'jpg'
                
                # PNG 检测
                if header.startswith(b'\x89PNG\r\n\x1a\n'):
                    return 'png'
                
                # WebP 检测
                if header.startswith(b'RIFF') and header[8:12] == b'WEBP':
                    return 'webp'
                
                # BMP 检测
                if header.startswith(b'BM'):
                    return 'bmp'
                
                # TIFF 检测
                if header.startswith(b'II*\x00') or header.startswith(b'MM\x00*'):
                    return 'tiff'
                
                return None
                
        except Exception as e:
            print(f"读取文件失败: {file_path} - {e}")
            return None

class ImageProcessor:
    """图片处理器"""
    
    def __init__(self, max_workers: int = 10):
        self.max_workers = max_workers
        self.detector = ImageFormatDetector()
        self.stats = {
            'processed': 0,
            'renamed': 0,
            'errors': 0,
            'skipped': 0
        }
    
    def get_file_extension(self, file_path: str) -> str:
        """获取文件扩展名（不包含点）"""
        return Path(file_path).suffix.lower().lstrip('.')
    
    async def process_file(self, file_path: Path) -> None:
        """
        处理单个文件
        
        Args:
            file_path: 文件路径
        """
        try:
            if not file_path.is_file():
                return
            
            current_extension = self.get_file_extension(str(file_path))
            
            # 只处理常见的图片扩展名
            image_extensions = {'jpg', 'jpeg', 'gif', 'png', 'webp', 'bmp', 'tiff'}
            if current_extension not in image_extensions:
                return
            
            self.stats['processed'] += 1
            
            # 检测真实格式
            actual_format = await self.detector.detect_format(str(file_path))
            if not actual_format:
                print(f"⚠️  无法检测格式: {file_path.name}")
                self.stats['skipped'] += 1
                return
            
            # 判断是否需要重命名
            needs_rename = (
                current_extension != actual_format and 
                not (current_extension == 'jpeg' and actual_format == 'jpg')
            )
            
            if needs_rename:
                # 构造新的文件名
                new_name = file_path.stem + f'.{actual_format}'
                new_path = file_path.parent / new_name
                
                # 检查目标文件是否已存在
                if new_path.exists():
                    print(f"⚠️  目标文件已存在，跳过重命名: {file_path.name} -> {new_name}")
                    self.stats['skipped'] += 1
                    return
                
                # 重命名文件
                file_path.rename(new_path)
                print(f"✅ 已重命名: {file_path.name} -> {new_name} ({current_extension} -> {actual_format})")
                self.stats['renamed'] += 1
            else:
                print(f"✓  格式正确: {file_path.name} ({actual_format})")
                
        except Exception as e:
            print(f"❌ 处理文件失败: {file_path.name} - {e}")
            self.stats['errors'] += 1
    
    async def process_directory(self, dir_path: str) -> Dict[str, int]:
        """
        处理目录中的所有图片文件
        
        Args:
            dir_path: 目录路径
            
        Returns:
            处理统计信息
        """
        try:
            path = Path(dir_path)
            if not path.exists() or not path.is_dir():
                raise ValueError(f"目录不存在或不是有效目录: {dir_path}")
            
            # 获取所有文件
            files = [f for f in path.iterdir() if f.is_file()]
            
            # 使用信号量控制并发数
            semaphore = asyncio.Semaphore(self.max_workers)
            
            async def process_with_semaphore(file_path):
                async with semaphore:
                    await self.process_file(file_path)
            
            # 并发处理所有文件
            await asyncio.gather(*[process_with_semaphore(f) for f in files])
            
        except Exception as e:
            print(f"处理目录失败: {dir_path} - {e}")
            self.stats['errors'] += 1
        
        return self.stats

async def main():
    """主函数"""
    # 获取命令行参数
    target_dir = sys.argv[1] if len(sys.argv) > 1 else './Ori'
    
    print(f"🚀 开始处理目录: {target_dir}")
    print("正在检测和修复图片文件格式...\n")
    
    start_time = time.time()
    
    # 创建处理器并处理目录
    processor = ImageProcessor(max_workers=10)
    stats = await processor.process_directory(target_dir)
    
    end_time = time.time()
    
    # 输出统计信息
    print(f"\n{'='*50}")
    print("📋 处理完成统计:")
    print(f"   处理文件数: {stats['processed']}")
    print(f"   重命名文件数: {stats['renamed']}")
    print(f"   跳过文件数: {stats['skipped']}")
    print(f"   错误文件数: {stats['errors']}")
    print(f"   用时: {end_time - start_time:.2f}秒")
    print(f"{'='*50}")

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n程序被用户中断")
        sys.exit(1)
    except Exception as e:
        print(f"程序执行失败: {e}")
        sys.exit(1)
